### PR TITLE
Add documentation for aggregation

### PR DESCRIPTION
**Pull request by kidrahahjo**
## Description

Add documentation for the newly introduced aggregation feature.

## Motivation and Context

This should be merged after https://github.com/glotzerlab/signac-flow/pull/464 gets merged into master.
## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-docs/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-docs/blob/master/ContributorAgreement.md).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-docs/blob/master/CONTRIBUTING.md#code-style) of this project.

### CODE REVIEW

File: docs/source/aggregation.rst [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+Sorting jobs for aggregation
+----------------------------
+
+**signac-flow** allows users to define the sorting order of jobs before creating the aggregates with the
+help of ``sort_by`` parameter and the sorting order can be defined with the help of ``sort_ascending`` parameter.
+
+.. code-block:: python
+
+    @aggregator.groupsof(2, sort_by='temperature', sort_ascending=False)
```
Comment (csadorf) [inline docs/source/aggregation.rst:102]: I would maybe suggest to rename `sort_ascending` to `sort_reverse` to be closer to the API of the [standard `sorted()` function](https://docs.python.org/3/library/functions.html#sorted).

Comment (csadorf) [reply docs/source/aggregation.rst:102]: @bdice What do you think?

File: docs/source/aggregation.rst [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        pass
+
+.. note::
+
+    In the above example, all the jobs will be sorted by the state point parameter ``temperature`` in descending
+    order and then be aggregated as groups of 2.
+
+Selecting jobs for aggregation
+------------------------------
+**signac-flow** allows users to selectively choose which jobs to pass into operation functions.
```
Comment (csadorf) [inline docs/source/aggregation.rst:114]: Missing empty line.

File: docs/source/aggregation.rst [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        pass
+
+
+.. _aggregate_id:
+
+Aggregate ID
+============
+
+Similar to the concept of a job id, an aggregate id is a unique hash identifying an aggregate of jobs.
+The aggregate id is sensitive to the order of the jobs in the aggregate
```
Comment (csadorf) [inline docs/source/aggregation.rst:130]: So what is the default order then? Should be documented here.

File: docs/source/aggregation.rst [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+Aggregation with FlowGroups
+===========================
+
+In order to associate aggregator object with a :py:class:`FlowGroup`, **signac-flow** provides a
+``aggregator_obj`` parameter in :meth:`~flow.FlowProject.make_group`. By default, no aggregation takes
+place for a :py:class:`FlowGroup`.
+
+.. note::
+
+    Currently, **signac-flow** only allows single :class:`~flow.aggregator` per group, i.e. all the operations present
```
Comment (csadorf) [inline docs/source/aggregation.rst:158]:
```suggestion
    Currently, **signac-flow** only allows single :class:`~flow.aggregator` per group, i.e., all the operations present
```

File: docs/source/aggregation.rst [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    in a :py:class:`FlowGroup` will be using a same :class:`~flow.aggregator` object.
```
Comment (csadorf) [inline docs/source/aggregation.rst:159]:
```suggestion
    in a :py:class:`FlowGroup` will be using the same :class:`~flow.aggregator` object.
```

File: docs/source/aggregation.rst [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+.. code-block:: python
+
+    # project.py
+    from flow import FlowProject, aggregator
+
+    class Project(FlowProject):
+        pass
+
+    group = Project.make_group('agg-group', aggregator_obj=aggregator())
```
Comment (csadorf) [inline docs/source/aggregation.rst:169]: I think the example would be easier to follow if you provided an implementation for such an aggregator class.

**Review by csadorf — CHANGES_REQUESTED** [review]
Sorry for taking so long on this. Here is my first pass!

**Comment by b-butler** [comment]
@kidrahahjo can you still work on this? If not (which is fine) we could assign someone else so that we can prioritize getting this updated on the documentation.

**Comment by bdice** [comment]
I spoke with @DomFijan and he expressed interest in helping as well.

**Comment by kidrahahjo** [comment]
I'll try to update this PR by addressing @csadorf 's reviews within a few hours. I don't have any problems with anyone else working on this. Please feel free to contribute.

File: docs/source/aggregation.rst [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+:class:`~flow.aggregator.groupsof` allows users to aggregate jobs by generating aggregates of a given size.
+
+.. code-block:: python
+
+    @aggregator.groupsof(2)
+    @Project.operation
+    def op4(job1, job2):
+        pass
+
+In the above example, the jobs will get aggregated in groups of 2 and hence, only two jobs will
```
[the remark this answers is not in the document]
Comment (kidrahahjo) [reply docs/source/aggregation.rst:86]: I think you're right. In this case, "up to two" will be more appropriate.

File: docs/source/aggregation.rst [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+be passed as parameters at once.
+
+.. note::
+
+    In case the number of jobs in the project is odd, there will be one aggregate containing only a single
+    job and hence users should be careful while defining the parameters for an *aggregate operation*.
```
[the remark this answers is not in the document]
Comment (kidrahahjo) [reply docs/source/aggregation.rst:92]: Since in the above example the parity of total jobs is not defined, I think it'll be more appropriate to include this in the above example itself

File: docs/source/aggregation.rst (hunk 1 above)
Comment (kidrahahjo) [reply docs/source/aggregation.rst:102]: @bdice this seems like an API change. What do you propose here? I think since we've not released aggregation yet, this can be done.

File: docs/source/aggregation.rst (hunk 6 above)
Comment (kidrahahjo) [reply docs/source/aggregation.rst:169]: @csadorf, I am not sure I completely follow what you say. Can you please clarify?

File: docs/source/aggregation.rst (hunk 3 above)
Comment (kidrahahjo) [reply docs/source/aggregation.rst]: Tried addressing here: https://github.com/glotzerlab/signac-docs/pull/122/files#diff-d053cd842dd6fe746188c4339762f7e336b763b94d513473d65900b9836a09f0R89

File: docs/source/aggregation.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (kidrahahjo) [reply docs/source/aggregation.rst]: Tried addressing here: https://github.com/glotzerlab/signac-docs/pull/122/files#diff-d053cd842dd6fe746188c4339762f7e336b763b94d513473d65900b9836a09f0R83

**Comment by DomFijan** [comment]
When I was trying out aggregation yesterday and it wasn't immediately obvious to me that functions passed to `@Project.post/pre`  need to now have a different input argument (`job` vs `*jobs`). If one wants to use aggregation to "embarrassingly parallelise" certain set of jobs (which I foresee will be a common application of this feature as bundling is often so problematic) the whole "workflow" of the operation one wishes to aggregate needs to be changed. This involves writing filters to filter out jobs which might have already been finished, rewriting pre/post condition functions for aggregated operation as outlined above, and finally splitting the MPI communicator for each job that one wishes to aggregate, and possibly some more... 
My question is - should this be metioned somewhere in the documentation? Perhaps not in a form of an example in this place, but in such a capacity that user is aware that some modifications are needed, and simply adding an aggregate decorator will not just work automagically in such cases? Perhaps a followup example somewhere else in the docs might be useful to demonstrate this capability to provide a more well-rounded example of usefulness of this feature.

File: docs/source/aggregation.rst [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+    @Project.operation
+    def op2(job):
+        pass
+
+    if __name__ == '__main__':
+        Project().main()
+
+By default, if :class:`~flow.aggregator` is used as a decorator, aggregate of all the jobs present in the project will be created.
+In the above example, ``op1`` can be referred to as an *aggregate operation* where all the jobs present in the project are passed as arbitraty arguments (or ``*args``) and ``op2`` is a *normal operation* where only a single job is passed as a parameter.
```
Comment (DomFijan) [inline docs/source/aggregation.rst:38]: I think mentioning the fact that all the functions (not sure if all, but definitely most?) used by the operation, such as functions of post and pre conditions, now require passing of `*jobs` instead of a `job`.

File: docs/source/aggregation.rst [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+    @group
+    @Project.operation
+    def op2(*jobs):
+        pass
+
+    if __name__ == '__main__':
+        Project().main()
+
+In the above example, when the group ``agg-group`` is executed, all the jobs in the project are passed as arbitrary arguments for ``op1`` and ``op2``. But if only ``op2`` is executed, only a single job is passed as a parameter.
```
Comment (Charlottez112) [inline docs/source/aggregation.rst:172]: This last sentence is also a bit confusing to me (probably because I'm still not very familiar with `aggregation`). Users might be wondering if a job is randomly chosen and passed in as a parameter. I assume you mean when used alone, `op2` operates on a single-job basis, but still iterates over all jobs? (Correct me if I'm wrong)

File: docs/source/aggregation.rst [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+    Users can also pass an aggregate id to the ``--job-id`` command-line flag provided by **signac-flow** in ``run``, ``submit``, and ``exec``.
+
+
+.. _aggregation_with_flow_groups:
+
+Aggregation with FlowGroups
+===========================
+
+In order to associate aggregator object with a :py:class:`FlowGroup`, **signac-flow** provides a ``group_aggregator`` parameter in :meth:`~flow.FlowProject.make_group`. By default, no aggregation takes place for a :py:class:`FlowGroup`.
```
Comment (Charlottez112) [inline docs/source/aggregation.rst:142]: For the last setence, do you mean that `FlowGroup` operates on all jobs and does not generate aggregates by default?

File: docs/source/aggregation.rst (hunk 1 above)
Comment (Charlottez112) [reply docs/source/aggregation.rst:93]: Although this is not a question for me, from a user perspective I guess `sort_ascending` is more descriptive. If we use `sort_reverse`, I guess we still need to point out that it follows the standard `sorted()` convention, assuming not everyone knows what `sorted` does exactly.

File: docs/source/aggregation.rst [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        pass
+
+    @Project.operation
+    def op2(job):
+        pass
+
+    if __name__ == '__main__':
+        Project().main()
+
+By default, if :class:`~flow.aggregator` is used as a decorator, aggregate of all the jobs present in the project will be created.
```
Comment (Charlottez112) [inline docs/source/aggregation.rst:37]: This is a bit ambiguous. 
```suggestion
If :class:`~flow.aggregator` is used with the default arguments, an aggregate of all the jobs present in the project will be created.
```

File: docs/source/aggregation.rst [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+This chapter provides information about the passing aggregates of jobs to operation functions.
+
+
+.. _aggregator_definition:
+
+aggregator
+==========
+
+An :class:`~flow.aggregator` is used as a decorator for an operation function which acts as an entry point to define the type of aggregation to perform while submitting or running the operation function.
```
Comment (Charlottez112) [inline docs/source/aggregation.rst:15]: Could you try to rephrase the definition? I guess we don't have to be limited to one sentence, and I don't quite follow which "entry point" you're referring to. Can we also use it on multiple operation functions?

**Comment by Charlottez112** [comment]
Thank you @kidrahahjo so much for the documentation! As someone who's not super familiar with `aggregation`, I found the examples very helpful. I just made a few comments on things that were not immediately clear to me. Could you also resolve the comments/ suggestions you've applied to the doc?

File: docs/source/aggregation.rst (hunk 11 above)
Comment (kidrahahjo) [reply docs/source/aggregation.rst:142]: Yes, since aggregation is implemented in a way that maintains backwards compatibility, the default behaviour of everything is similar to "pre-aggregation" phase of signac-flow.

File: docs/source/aggregation.rst (hunk 10 above)
Comment (kidrahahjo) [reply docs/source/aggregation.rst:172]: In the above example, `op1` have a default aggregator associated with it, `op2` doesn't have any aggregator associated with it, and the group `agg-group` (which contains both `op1` and `op2`) has a default aggregaot associated with it.
Here is the detailed explaination of how these will get executed

1. `python project.py -o agg-group` -> all jobs will get aggregated. So `op1` and `op2` will run once and the argument passed will be the aggregate of all jobs.
2. `python project.py -o op1` -> all jobs will get aggregated and `op1` will run once and the argument passed will be the aggregate of all jobs.
3. `python project.py -o op2` -> No aggregation will take place. Every job in the project will be passed as an argument individually to `op2`.

Does this explaination make the example more clear?

File: docs/source/aggregation.rst (hunk 13 above)
Comment (kidrahahjo) [reply docs/source/aggregation.rst]: Tried resolving here: https://github.com/glotzerlab/signac-docs/pull/122/files#diff-d053cd842dd6fe746188c4339762f7e336b763b94d513473d65900b9836a09f0R15

File: docs/source/aggregation.rst (hunk 10 above)
Comment (kidrahahjo) [reply docs/source/aggregation.rst]: Tried making it more clear here: https://github.com/glotzerlab/signac-docs/pull/122/files#diff-d053cd842dd6fe746188c4339762f7e336b763b94d513473d65900b9836a09f0R176

File: docs/source/aggregation.rst (hunk 9 above)
Comment (kidrahahjo) [reply docs/source/aggregation.rst]: Tried resolving here: https://github.com/glotzerlab/signac-docs/pull/122/files#diff-d053cd842dd6fe746188c4339762f7e336b763b94d513473d65900b9836a09f0R41

File: docs/source/aggregation.rst (hunk 10 above)
Comment (Charlottez112) [reply docs/source/aggregation.rst]: Your explanation here is clear, but still in the documentation it's a bit confusing. What I was trying to get at is that the sentence:

> only a single job is passed as a parameter

is not clear enough. I think this is also related to @DomFijan's comment on adding docs for `job` vs `*jobs`

File: docs/source/aggregation.rst [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+.. code-block:: python
+
+    @aggregator.groupsof(2, sort_by='temperature', sort_ascending=False)
+    @Project.operation
+    def op5(job1, job2):
+        pass
+
+.. note::
+
+    In the above example, all the jobs will be sorted by the state point parameter ``temperature`` in descending order and then be aggregated as groups of 2.
```
Comment (bdice) [inline docs/source/aggregation.rst:102]:
```suggestion
    In the above example, all the jobs will be sorted by the state point parameter ``"temperature"`` in descending order and then be aggregated as groups of 2.
```

File: docs/source/aggregation.rst (hunk 1 above)
Comment (bdice) [reply docs/source/aggregation.rst:95]:
```suggestion
    @aggregator.groupsof(2, sort_by="temperature", sort_ascending=False)
```

File: docs/source/aggregation.rst [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+----------------------------
+
+**signac-flow** allows users to define the sorting order of jobs before creating the aggregates with the help of ``sort_by`` parameter and the sorting order can be defined with the help of ``sort_ascending`` parameter.
+By default, when no ``sort_by`` parameter is specified, the order of the jobs will be decided by the order in which the jobs are iterated in a **signac** project.
+
+.. code-block:: python
+
+    @aggregator.groupsof(2, sort_by='temperature', sort_ascending=False)
+    @Project.operation
+    def op5(job1, job2):
```
Comment (bdice) [inline docs/source/aggregation.rst:97]: Do we need to use `*jobs` or `job2=None` to support a final aggregate with one job?
(I also worry that showing examples with `job1, job2` will be more confusing than `*jobs`.)

File: docs/source/aggregation.rst [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+Selecting jobs for aggregation
+------------------------------
+
+**signac-flow** allows users to selectively choose which jobs to pass into operation functions.
+
+.. code-block:: python
+
+    @aggregator(select=lambda job: job.sp.temperature > 0)
+    @Project.operation
+    def op6(job1, job2):
```
Comment (bdice) [inline docs/source/aggregation.rst:113]: We don't expect this to return two jobs.
```suggestion
    def op6(*jobs):
```

File: docs/source/aggregation.rst [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+
+Aggregation with FlowGroups
+===========================
+
+In order to associate aggregator object with a :py:class:`FlowGroup`, **signac-flow** provides a ``group_aggregator`` parameter in :meth:`~flow.FlowProject.make_group`.
+By default, no aggregation takes place for a :py:class:`FlowGroup`.
+
+.. note::
+
+    Currently, **signac-flow** only allows single :class:`~flow.aggregator` per group, i.e., all the operations present in a :py:class:`FlowGroup` will be using the same :class:`~flow.aggregator` object.
```
Comment (bdice) [inline docs/source/aggregation.rst:149]:
```suggestion
    All the operations in a :py:class:`FlowGroup` will use the same :class:`~flow.aggregator` object provided to the group's ``group_aggregator`` parameter.
```

File: docs/source/aggregation.rst [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+
+    @group
+    @Project.operation
+    def op2(*jobs):
+        pass
+
+    if __name__ == '__main__':
+        Project().main()
+
+In the above example, when the group ``agg-group`` is executed using ``python project.py run -o agg-group``, all the jobs in the project are passed as arbitrary arguments for both, ``op1`` and ``op2``.
```
Comment (bdice) [inline docs/source/aggregation.rst:175]:
```suggestion
In the above example, when the group ``agg-group`` is executed using ``python project.py run -o agg-group``, all the jobs in the project are passed as positional arguments for both ``op1`` and ``op2``.
```

File: docs/source/aggregation.rst [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+But if ``op2`` is executed using ``python project.py run -o op2``, only a single job is passed as an argument because no :class:`~flow.aggregator` is associated with the operation function ``op2``.
```
Comment (bdice) [inline docs/source/aggregation.rst:176]:
```suggestion
If ``op2`` is executed using ``python project.py run -o op2``, only a single job is passed as an argument because no :class:`~flow.aggregator` is associated with the operation function ``op2``.
```

**Review by bdice — CHANGES_REQUESTED** [review]
Comments below. Overall this is great and covers the right topics! Just need to clean it up a bit. I think we should show fewer examples of `def my_op(job1, job2)` and replace them with `def my_op(*jobs)`. I think that is a better practice to demonstrate for most users' needs.

**Comment by csadorf** [comment]
@kidrahahjo Please re-request my review after addressing @bdice 's comments. Thank you.

File: docs/source/aggregation.rst [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+In the above example, the jobs will get aggregated in groups of 2 and hence, up to two jobs will be passed as arguments at once.
+
+.. note::
+
+    In case the number of jobs in the project is odd, there will be one aggregate containing only a single job and hence users should be careful while passing non-default arguments in an *aggregate operation*.
+
+Sorting jobs for aggregation
+----------------------------
+
+**signac-flow** allows users to define the sorting order of jobs before creating the aggregates with the help of ``sort_by`` parameter and the sorting order can be defined with the help of ``sort_ascending`` parameter.
```
[the remark this answers is not in the document]
Comment (kidrahahjo) [reply docs/source/aggregation.rst:90]: Writing "sorting order" in both the lines looks confusing to me. Can we make it like, _"Aggregators allow users to sort jobs before creating an aggregate with the `sort_by` parameter. The sorting order can be defined with the `sort_ascending` parameter."_

[the remark this answers is not in the document]
Comment (kidrahahjo) [reply docs/source/aggregation.rst:90]: @bdic I'll commit the suggested changes for now, we can reiterate on this.

[the remark this answers is not in the document]
Comment (bdice) [reply docs/source/aggregation.rst:90]: Yes, your suggestion is good!

File: docs/source/aggregation.rst (hunk 15 above)
Comment (kidrahahjo) [reply docs/source/aggregation.rst:97]: Since we've made our point of using non default arguments carefully, I think we should go with `*jobs` here.

File: docs/source/aggregation.rst [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+.. _types_of_aggregation:
+
+Types of Aggregation
+====================
+
+Currently, **signac-flow** allows users to aggregate jobs by:
+
+- Grouping them on state point key, an iterable of state point keys whose values define the groupings, or an arbitrary callable of :class:`~signac.contrib.job.Job`.
+- Generating aggregates of a given size.
+- Using custom aggregator function when greater flexibility is needed.
```
Comment (csadorf) [inline docs/source/aggregation.rst:55]:
```suggestion
Currently, **signac-flow** allows users to aggregate jobs in the following way:

- _All jobs_: All of the project's jobs are passed to the operation function.
- _Group by state point key(s)_: The aggregates are grouped by one or more state point keys.
- _Group by arbitrary key-function_: The aggregates are grouped by keys determined by a key-function that expects an instance of :class:`~.signac.contrib.job.Job` and return the grouping key.
- Grouping into aggregates of a specific size.
- Using a completely custom aggregator function when even greater flexibility is needed.
```

File: docs/source/aggregation.rst [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+
+    @Project.operation
+    def op2(job):
+        pass
+
+    if __name__ == '__main__':
+        Project().main()
+
+If :class:`~flow.aggregator` is used with the default arguments, it will create a single aggregate containing all the jobs present in the project.
+In the example above, ``op1`` is an *aggregate operation* where all the jobs present in the project are passed as a variable number of positional arguments (via ``*jobs``), while ``op2`` is a normal operation where only a single job is passed as an argument.
```
Comment (csadorf) [inline docs/source/aggregation.rst:40]:
```suggestion
In the example above, ``op1`` is an *aggregate operation* where all the jobs present in the project are passed as a variable number of positional arguments (via ``*jobs``), while ``op2`` is an operation where only a single job is passed as an argument.

.. tip:

    The concept of aggregation might be easier to understand if one realizes that "normal" operation functions are equivalent to aggregate-operations-functions with an aggregate group size of one.
```
I don't like the distinction between "aggregate" and "normal" operation functions as it implies that aggregate operations are somehow not normal. However, that's not the case. Every operation function is an aggregate operation function, just that "normal operations" are in the limit of aggregate size one. I already explained previously in the context of documenting "groups": It is important to not introduce special cases, but instead try to relate the concepts to each other in a logical way.

- For a user it is easier to make the transfer from "normal" operations to operation groups, when they realize that each operation constitutes an operation group with a single entry.
- Similarly, I believe that a user will have an easier time to understand the aggregate-concept, when they realize that "normal operation functions" are aggregate operation functions of size 1.

In summary: in my view all operation functions are in principle the same, just by default they form a group of size 1 and an aggregate of size 1.

File: docs/source/aggregation.rst [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+    def op4(job1, job2=None):
+        pass
+
+In the above example, the jobs will get aggregated in groups of 2 and hence, up to two jobs will be passed as arguments at once.
+
+.. note::
+
+    In case the number of jobs in the project in this example is odd, there will be one aggregate containing only a single job.
+    In general, the last aggregate from :class:`~flow.aggregator.groupsof` will contain the remaining jobs if the aggregate size does not evenly divide the number of jobs in the project.
+    Users should be careful while passing non-default arguments in an *aggregate operation*.
```
Comment (csadorf) [inline docs/source/aggregation.rst:90]:
```suggestion
    If a remainder is expected and valid, users should make sure that the operation function can be called with the reduced number of arguments (e.g. by using `*jobs*` or providing default arguments as shown above).
```

File: docs/source/aggregation.rst (hunk 1 above)
Comment (csadorf) [reply docs/source/aggregation.rst:101]: Well, the feature is released now, so we just go with whatever the API is now, right?

**Review by csadorf — CHANGES_REQUESTED** [review]

File: docs/source/aggregation.rst (hunk 22 above)
Comment (bdice) [reply docs/source/aggregation.rst:40]: I agree with this approach. I will leave this comment thread open until I have a chance to review again and see if there are other instances of this distinction. If this is the only place, we can just commit the suggested change.

File: docs/source/aggregation.rst [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+    def op4(job1, job2=None):
+        pass
+
+In the above example, the jobs will get aggregated in groups of 2 and hence, up to two jobs will be passed as arguments at once.
+
+.. note::
+
+    In case the number of jobs in the project in this example is odd, there will be one aggregate containing only a single job.
+    In general, the last aggregate from :class:`~flow.aggregator.groupsof` will contain the remaining jobs if the aggregate size does not evenly divide the number of jobs in the project.
+    If a remainder is expected and valid, users should make sure that the operation function can be called with the reduced number of arguments (e.g. by using `*jobs*` or providing default arguments as shown above).
```
Comment (bdice) [inline docs/source/aggregation.rst:92]:
```suggestion
    If a remainder is expected and valid, users should make sure that the operation function can be called with the reduced number of arguments (e.g. by using ``*jobs`` or providing default arguments as shown above).
```

File: docs/source/aggregation.rst [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+- *All jobs*: All of the project's jobs are passed to the operation function.
+- *Group by state point key(s)*: The aggregates are grouped by one or more state point keys.
+- *Group by arbitrary key function*: The aggregates are grouped by keys determined by a key-function that expects an instance of :class:`~.signac.contrib.job.Job` and return the grouping key.
+- Grouping into aggregates of a specific size.
+- Using a completely custom aggregator function when even greater flexibility is needed.
+
+Group By
+--------
+
+:class:`~flow.aggregator.groupby` allows users to aggregate jobs by grouping them on state point key, an iterable of state point keys whose values define the groupings, or an arbitrary callable of :class:`~signac.contrib.job.Job`.
```
Comment (bdice) [inline docs/source/aggregation.rst:62]:
```suggestion
:class:`~flow.aggregator.groupby` allows users to aggregate jobs by grouping them by a state point key, an iterable of state point keys whose values define the groupings, or an arbitrary callable of :class:`~signac.contrib.job.Job`.
```

File: docs/source/aggregation.rst [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+        pass
+
+.. note::
+
+    In the above example, all the jobs will be sorted by the state point parameter ``"temperature"`` in descending order and then be aggregated as groups of 2.
+
+Selecting jobs for aggregation
+------------------------------
+
+**signac-flow** allows users to selectively choose which jobs to pass into operation functions.
```
Comment (bdice) [inline docs/source/aggregation.rst:115]:
```suggestion
**signac-flow** allows users to selectively choose which jobs to pass into operation functions.
This can be used to generate aggregates from only the selected jobs, excluding any jobs that do not meet the selection criteria.
```

File: docs/source/aggregation.rst [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+
+.. code-block:: python
+
+    # project.py
+    from flow import FlowProject, aggregator
+
+    class Project(FlowProject):
+        pass
+
+    group = Project.make_group('agg-group', group_aggregator=aggregator())
```
Comment (bdice) [inline docs/source/aggregation.rst:168]:
```suggestion
    group = Project.make_group("agg-group", group_aggregator=aggregator())
```

File: docs/source/aggregation.rst [hunk 28]
```diff
@@ -0,0 +1,185 @@
...
+    @Project.operation
+    def op1(*jobs):
+        pass
+
+    @group
+    @Project.operation
+    def op2(*jobs):
+        pass
+
+    if __name__ == '__main__':
```
Comment (bdice) [inline docs/source/aggregation.rst:181]:
```suggestion
    if __name__ == "__main__":
```

File: docs/source/aggregation.rst [hunk 29]
```diff
@@ -0,0 +1,185 @@
...
+        Project().main()
+
+In the above example, when the group ``agg-group`` is executed using ``python project.py run -o agg-group``, all the jobs in the project are passed as positional arguments for both ``op1`` and ``op2``.
+If ``op2`` is executed using ``python project.py run -o op2``, only a single job is passed as an argument because no :class:`~flow.aggregator` is associated with the operation function ``op2``.
```
Comment (bdice) [inline docs/source/aggregation.rst:185]:
```suggestion
If ``op1`` is executed using ``python project.py run -o op1``, all the jobs in the project are passed as positional arguments because a :class:`~flow.aggregator` is associated with the operation function ``op1`` (separately from the aggregator used for ``agg-group``).
If ``op2`` is executed using ``python project.py run -o op2``, only a single job is passed as an argument because no :class:`~flow.aggregator` is associated with the operation function ``op2``.
```

File: docs/source/aggregation.rst [hunk 30]
```diff
@@ -0,0 +1,186 @@
...
+    @aggregator()
+    @Project.operation
+    def op1(*jobs):
+        print("Number of jobs in aggregate:", len(jobs))
+
+    @Project.operation
+    def op2(job):
+        pass
+
+    if __name__ == '__main__':
```
Comment (bdice) [inline docs/source/aggregation.rst:36]:
```suggestion
    if __name__ == "__main__":
```

File: docs/source/aggregation.rst (hunk 22 above)
Comment (bdice) [reply docs/source/aggregation.rst:40]: @csadorf Do you want to make an section that explicitly describes your point about all operations acting with aggregation in the limit of one job? I think that could strengthen the docs.

File: docs/source/aggregation.rst [hunk 31]
```diff
@@ -0,0 +1,186 @@
...
+        pass
+
+In the above example, the jobs will get aggregated in groups of 2 and hence, up to two jobs will be passed as arguments at once.
+
+.. note::
+
+    In case the number of jobs in the project in this example is odd, there will be one aggregate containing only a single job.
+    In general, the last aggregate from :class:`~flow.aggregator.groupsof` will contain the remaining jobs if the aggregate size does not evenly divide the number of jobs in the project.
+    If a remainder is expected and valid, users should make sure that the operation function can be called with the reduced number of arguments (e.g. by using ``*jobs`` or providing default arguments as shown above).
+
```
Comment (bdice) [inline docs/source/aggregation.rst:93]: We could add more examples for some or all of the following:
- *Group by state point keys*: The aggregates are grouped by multiple state point keys.
- *Group by arbitrary key function*: The aggregates are grouped by keys determined by a key-function that expects an instance of :class:`~.signac.contrib.job.Job` and return the grouping key.
- Using a completely custom aggregator function when even greater flexibility is needed.

File: docs/source/aggregation.rst [hunk 32]
```diff
@@ -0,0 +1,186 @@
...
+    @aggregator(select=lambda job: job.sp.temperature > 0)
+    @Project.operation
+    def op6(*jobs):
+        pass
+
+
+.. _aggregate_id:
+
+Aggregate ID
```
Comment (bdice) [inline docs/source/aggregation.rst:128]: Hmm. I don't think we capitalize "id" anywhere else in the docs so it feels weird to do that here. We aren't using title case consistently in the section headers of this page (e.g. "Selecting jobs for aggregation" is not title case, while "Groups Of" is using title case). On closer inspection, we're inconsistent about capitalization in section headings across the entirety of the signac documentation. I don't know what to do here.

File: docs/source/aggregation.rst [hunk 33]
```diff
@@ -0,0 +1,186 @@
...
+
+Similar to the concept of a job id, an aggregate id is a unique hash identifying an aggregate of jobs.
+The aggregate id is sensitive to the order of the jobs in the aggregate.
+
+
+.. note::
+
+    The id of an aggregate containing one job is that job's id.
+
+In order to distinguish between aggregate id and a job id, for an aggregate of more than one job the aggregate id of that aggregate will always have a prefix ``agg-``.
```
Comment (bdice) [inline docs/source/aggregation.rst:139]:
```suggestion
In order to distinguish between an aggregate id and a job id, the id of aggregates with more than one job will always have a prefix ``agg-``.
```

File: docs/source/aggregation.rst [hunk 34]
```diff
@@ -0,0 +1,186 @@
...
+
+    Users can also pass an aggregate id to the ``--job-id`` command-line flag provided by **signac-flow** in ``run``, ``submit``, and ``exec``.
+
+
+.. _aggregation_with_flow_groups:
+
+Aggregation with FlowGroups
+===========================
+
+In order to associate aggregator object with a :py:class:`FlowGroup`, **signac-flow** provides a ``group_aggregator`` parameter in :meth:`~flow.FlowProject.make_group`.
```
Comment (bdice) [inline docs/source/aggregation.rst:153]:
```suggestion
In order to associate an aggregator object with a :py:class:`FlowGroup`, **signac-flow** provides a ``group_aggregator`` parameter in :meth:`~flow.FlowProject.make_group`.
```

**Review by bdice — APPROVED** [review]
I think this PR is mostly complete, so I am approving it. All of my remaining suggestions could be addressed in a separate PR if desired.

**Review by Charlottez112 — APPROVED** [review]
Referring to Python unpacking is very helpful!

File: docs/source/aggregation.rst (hunk 22 above)
Comment (bdice) [reply docs/source/aggregation.rst:40]: > Do you want to make an section that explicitly describes your point about all operations acting with aggregation in the limit of one job?

I re-read the original comment. The suggestion posted by @csadorf includes exactly the language that I was hoping to add. Somehow I missed that the first time. 😅

File: docs/source/aggregation.rst (hunk 31 above)
Comment (bdice) [reply docs/source/aggregation.rst:98]: I created a new issue for this.

File: docs/source/aggregation.rst (hunk 32 above)
Comment (bdice) [reply docs/source/aggregation.rst:133]: Created #147 to address this.

**Comment by bdice** [comment]
I created #146, #147 to address my remaining comments on this PR. With two approvals, I think it's ready to merge. @csadorf I think we've addressed your comments sufficiently, but feel free to follow up if you wanted to make another round of changes. Thanks for your work on this, @kidrahahjo. 👍

**Comment by csadorf** [comment]
Thx a lot @kidrahahjo !!
